src/nvim/tui/tui.c

```diff
@@ -287,10 +288,30 @@ void tui_enable_extended_underline(TUIData *tui)
 static void tui_query_kitty_keyboard(TUIData *tui)
   FUNC_ATTR_NONNULL_ALL
 {
-  tui->input.waiting_for_kkp_response = true;
+  // Set the key encoding to the fallback value when the primary device
+  // attributes (DA1) response is received. If the response to the kitty query
+  // is received first, this function won't be called.
+  tui->input.callbacks.primary_device_attr = tui_set_key_encoding;
   out(tui, S_LEN("\x1b[?u\x1b[c"));
 }
 
+static void tui_reset_kitty_keyboard(TUIData *tui)
+{
+  if (tui->input.callbacks.primary_device_attr) {
+    // We've been called after receiving a DA1 response. Nothing to do.
+    return;
+  }
+
+  // Send a DA1 request after disabling key events. When we receive the
+  // response, we know there are no more key events to receive.
+  tui->input.callbacks.primary_device_attr = tui_reset_kitty_keyboard;
+  out(tui, S_LEN("\x1b[<u\x1b[c"));
+
+  // Immediately flush the buffer and wait for the DA1 response.
+  flush_buf(tui);
+  LOOP_PROCESS_EVENTS_UNTIL(tui->loop, NULL, 500, tui->input.callbacks.primary_device_attr == NULL);
```

Comment:
This doesn't work when suspending (Ctrl-Z) because the main event loop runs recursively.